temporal-sdk/src/main/java/io/temporal/internal/statemachines/NexusOperationStateMachine.java

```diff
@@ -0,0 +1,269 @@
+  private void notifyCompleted() {
+    notifyStarted();
+    NexusOperationCompletedEventAttributes attributes =
+        currentEvent.getNexusOperationCompletedEventAttributes();
+    completionCallback.apply(Optional.of(attributes.getResult()), null);
+  }
+
+  private void notifyFailed() {
+    notifyStarted();
+    NexusOperationFailedEventAttributes attributes =
+        currentEvent.getNexusOperationFailedEventAttributes();
+    completionCallback.apply(Optional.empty(), attributes.getFailure());
+  }
+
+  private void notifyCanceled() {
+    notifyStarted();
+    NexusOperationCanceledEventAttributes attributes =
+        currentEvent.getNexusOperationCanceledEventAttributes();
+    completionCallback.apply(Optional.empty(), attributes.getFailure());
+  }
+
+  private void notifyTimedOut() {
+    notifyStarted();
+    NexusOperationTimedOutEventAttributes attributes =
+        currentEvent.getNexusOperationTimedOutEventAttributes();
+    completionCallback.apply(Optional.empty(), attributes.getFailure());
+  }
+
+  /**
+   * @param attributes attributes used to schedule the nexus operation
+   * @param startedCallback invoked when the Nexus operation start
+   * @param completionCallback invoked when Nexus operation completes
+   * @param commandSink sink to send commands
+   * @return an instance of NexusOperationStateMachine
+   */
+  public static NexusOperationStateMachine newInstance(
+      ScheduleNexusOperationCommandAttributes attributes,
+      Functions.Proc2<Optional<String>, Failure> startedCallback,
+      Functions.Proc2<Optional<Payload>, Failure> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    return new NexusOperationStateMachine(
+        attributes, startedCallback, completionCallback, commandSink, stateMachineSink);
+  }
+
+  private NexusOperationStateMachine(
+      ScheduleNexusOperationCommandAttributes attributes,
+      Functions.Proc2<Optional<String>, Failure> startedCallback,
+      Functions.Proc2<Optional<Payload>, Failure> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    super(STATE_MACHINE_DEFINITION, commandSink, stateMachineSink);
+    this.scheduleAttributes = attributes;
+    this.operation = attributes.getOperation();
+    this.service = attributes.getService();
+    this.endpoint = attributes.getEndpoint();
+    this.startedCallback = startedCallback;
+    this.completionCallback = completionCallback;
+    explicitEvent(ExplicitEvent.SCHEDULE);
+  }
+
+  public void createScheduleNexusTaskCommand() {
+    addCommand(
+        Command.newBuilder()
+            .setCommandType(CommandType.COMMAND_TYPE_SCHEDULE_NEXUS_OPERATION)
+            .setScheduleNexusOperationCommandAttributes(scheduleAttributes)
+            .build());
+    scheduleAttributes = null; // avoiding retaining large input for the duration of the operation
```

Comment:
This is great, can you confirm that we never hold on to operation results in memory post completion?

Reply:
We don't, the result is just passed through the callback which sets the result on the promise. If users are holding a reference to the promise then the decoded operation results would be held of course.